Add endpoint for dynamic type schema

ndoschek opened the pull request:
Add endpoint for dynamic type schema
- Add new endpoint for retrieving a dynamically created type JSON schema
- Update ModelServerClient

Add endpoint for JsonForms UI schemas (was PR #39)
- Add endpoint for retrieving UI schemas from example workspace
- Add UI schemas to example workspace
- Update ModelServerClient

Add endpoint for fetching single model elements
- Fetch model elements via element id (uri fragment) or element name (if feature exists)

Please also see https://github.com/eclipse-emfcloud/emfcloud-modelserver-theia/pull/12

martin-fleck-at commented on bundles/org.eclipse.emfcloud.modelserver.client/src/org/eclipse/emfcloud/modelserver/client/ModelServerClient.java:
```diff
@@ -137,6 +137,44 @@ public void close() {
    }
 
+   @Override
+   public CompletableFuture<Response<EObject>> getModelElementById(final String modelUri, final String elementid,
```
Please add a test case for this in ModelServerClientTest.

martin-fleck-at commented on bundles/org.eclipse.emfcloud.modelserver.client/src/org/eclipse/emfcloud/modelserver/client/ModelServerClient.java:
```diff
@@ -137,6 +137,44 @@ public void close() {
+   }
+
+   @Override
+   public CompletableFuture<Response<EObject>> getModelElementByName(final String modelUri, final String elementname,
```
Please add a test case for this in ModelServerClientTest.

martin-fleck-at commented on bundles/org.eclipse.emfcloud.modelserver.client/src/org/eclipse/emfcloud/modelserver/client/ModelServerClient.java:
```diff
@@ -263,6 +301,34 @@ private boolean isSupportedFormat(final String format) {
    }
 
+   @Override
+   public CompletableFuture<Response<String>> getTypeSchema(final String modelUri) {
```
Please add a test case for this in ModelServerClientTest.

martin-fleck-at commented on bundles/org.eclipse.emfcloud.modelserver.client/src/org/eclipse/emfcloud/modelserver/client/ModelServerClient.java:
```diff
@@ -263,6 +301,34 @@ private boolean isSupportedFormat(final String format) {
+   }
+
+   @Override
+   public CompletableFuture<Response<String>> getUISchema(final String schemaname) {
```
Please add a test case for this in ModelServerClientTest.

martin-fleck-at commented on bundles/org.eclipse.emfcloud.modelserver.emf/src/org/eclipse/emfcloud/modelserver/emf/common/ModelController.java:
```diff
@@ -119,6 +119,36 @@ public void getOne(final Context ctx, final String modeluri) {
          () -> handleError(ctx, 404, "Model '" + modeluri + "' not found!"));
    }
 
+   public void getModelElement(final Context ctx, final String modeluri, final String elementid) {
```
Should this also match the 'getModelElementById' naming scheme? Could you please add a test case for this in ModelControllerTest.

ndoschek replied:
Yes I agree, this should be consistent :+1:

martin-fleck-at commented on bundles/org.eclipse.emfcloud.modelserver.emf/src/org/eclipse/emfcloud/modelserver/emf/common/ModelController.java:
```diff
@@ -119,6 +119,36 @@ public void getOne(final Context ctx, final String modeluri) {
+            "Element by elementid '" + elementid + "' of model '" + modeluri + "' not found!"));
+   }
+
+   public void getModelElementByName(final Context ctx, final String modeluri, final String elementname) {
```
Please add a test case for this in ModelControllerTest.

martin-fleck-at commented on bundles/org.eclipse.emfcloud.modelserver.emf/src/org/eclipse/emfcloud/modelserver/emf/common/ModelRepository.java:
```diff
@@ -108,6 +110,37 @@ boolean hasModel(final String modeluri) {
    }
 
+   @SuppressWarnings("checkstyle:IllegalCatch")
+   public Optional<EObject> getModelElementById(final String modeluri, final String elementid) {
```
Does it make sense to also add a test for this in ModelRepositoryTest?

martin-fleck-at commented on bundles/org.eclipse.emfcloud.modelserver.emf/src/org/eclipse/emfcloud/modelserver/emf/common/ModelRepository.java:
```diff
@@ -108,6 +110,37 @@ boolean hasModel(final String modeluri) {
+         });
+   }
+
+   public Optional<EObject> getModelElementByName(final String modeluri, final String elementname) {
```
Does it make sense to also add a test for this in ModelRepositoryTest?

Also this method will only return the FIRST element that matches the given name. That is fine by me but it should make sure that this is maybe documented somewhere. Afaik there is no real API documentation right now but if I am wrong please make sure that this behavior is documented.

ndoschek replied:
Yes this is a good point. I updated the README, where the current ModelServerClientAPI as well as the HTTP endpoints are documented. I added therefore a hint that it returns the first machting element.
Also I added a test in ModelRepositoryTest.

martin-fleck-at commented on bundles/org.eclipse.emfcloud.modelserver.emf/src/org/eclipse/emfcloud/modelserver/emf/common/SchemaController.java:
```diff
@@ -21,11 +21,30 @@
+   @Inject
+   private SchemaRepository schemaRepository;
+
    public void getSchema(final Context ctx, final String modeluri) {
```
It seems to me that the 'getTypeSchema' is more or less replacing the 'getSchema' method as the 'getSchema' method does not provide any useful information. If that is the case we should discuss whether it can be removed. This definitely does not have to happen as part of this change but maybe we can create a ticket for that if you agree.

ndoschek replied:
Yes I agree to remove it. As I had changes in this area anyways, I removed this method and all its occurrences.

martin-fleck-at commented on bundles/org.eclipse.emfcloud.modelserver.emf/src/org/eclipse/emfcloud/modelserver/emf/common/SchemaController.java:
```diff
@@ -21,11 +21,30 @@
+            .json(JsonResponse.error(String.format("Type schema for '%s' not found!", ctx.queryParam("modeluri")))));
+   }
+
+   public void getJsonFormsUISchema(final Context ctx, final String schemaname) {
```
Why does this method name have the specific 'JsonForms' in the name? Just wondering.

ndoschek replied:
I agree, this prefix can be removed as it does not add any value to the readability of the method's name ;)

martin-fleck-at commented on bundles/org.eclipse.emfcloud.modelserver.emf/src/org/eclipse/emfcloud/modelserver/emf/common/SchemaRepository.java:
```diff
@@ -0,0 +1,49 @@
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.google.inject.Inject;
+
+public class SchemaRepository {
```
Should this class also be bound as a Singleton in the ModelServerModule? Currently it is only injected once and only in another singleton. Furthermore it does not hold any state besides the schema repository path. So I am opening this up for discussion, what do you think?

ndoschek replied:
That's a good point, thanks for noticing! I bound it as Singleton now. Also on creating other unit tests this came up as an issue and therefore makes a lot of sense.

martin-fleck-at commented on bundles/org.eclipse.emfcloud.modelserver.emf/src/org/eclipse/emfcloud/modelserver/emf/configuration/ServerConfiguration.java:
```diff
@@ -48,6 +48,8 @@ public void setWorkspaceRoot(final String workspaceRoot) {
          .ifPresent(uri -> this.workspaceRootURI = uri);
    }
 
+   public String getWorkspaceUISchemaPath() { return workspaceRootURI.toFileString().concat(".ui-schemas/"); }
```
This is fine for now but I believe we should discuss more flexibility in the future if needed, e.g., at least having a setter for that so that we can read it from the command line or change it some other way. My main line of thinking is that we probably do not always want to have that on the workspace level.

ndoschek replied:
You're right, I used the same logic now to set the ui schema folder path as for the workspace root and detached it from the workspace root. It is now possible to set any valid file path as ui schema folder via the command line argument `-u`.

ndoschek commented:
Hi Martin, thanks for your great feedback on my changes. I added another two commits, the first mainly addresses your comments, and the second addresses the missing UnitTests.
